perf(router): optimize model lookups with O(1) index maps

AlexsanderHamir opened the pull request:
## Title

perf(router): optimize model lookups with O(1) index maps

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Performance Improvement

## Changes

- Some parts of the router were still performing lookups using lists instead of the available indexes, which is inefficient.

## Test Results

<img width="1425" height="195" alt="Screenshot 2025-10-15 at 3 19 30 PM" src="https://github.com/user-attachments/assets/f415a4b6-2a03-4bd6-ab19-b682198c1b0b" />

- Failures seem unrelated.

AlexsanderHamir committed:
perf(router): optimize model lookups with O(1) index maps and standardize timing

- Use model_id_to_deployment_index_map and model_name_to_deployment_indices for O(1) lookups in get_model_info, get_deployment_by_model_group_name, and get_model_ids

ishaan-jaff reviewed (changes requested):
missing some type of test to validate this continues using O(1) lookups for this

AlexsanderHamir committed:
test: add static analysis to prevent O(n) linear scans in router

Add AST-based test to detect 'for ... in self.model_list' anti-pattern.
Enforces use of index maps (model_id_to_deployment_index_map and
model_name_to_deployment_indices) for O(1) lookups instead of O(n) iteration.

AlexsanderHamir committed:
Merge branch 'litellm_october_alexsander_stanging' into litellm_router_index_change
